aeneas/src/mach/GlobalRegAlloc.v3

```diff
@@ -398,40 +539,66 @@ class GlobalRegAlloc(regSet: MachRegSet, codegen: SsaMachGen) {
 				var block = Operand.Label.!(cur.operands[0]).block;
 				computeLiveoutForBlock(block);
 				// Variables may be interfered with uses in different succ of the block.
-				for (i < vars.length) if (liveness[liveoutRow, i]) addInterfere(vars[i]);
+				for (i < vars.length) if (liveness[liveoutRow, i]) addInterfere(vars[i], i);
 				curWeight = blockWeight[block.mark];
 				cur = prev;
 				continue;
 			}
+			if (opcode == ArchInstrs.ARCH_BLOCK) {
+				var block = Operand.Label.!(cur.operands[0]).block;
+				liveness.copy(block.mark, liveoutRow);
+			}
 
 			var operands = cur.operands;
 			for (i < operands.length) match (operands[i]) {
 				Def(vreg, constraint) => {
+					// ignore `Def` introduced by previous spill or `insertLoadToStack`
+					if (vreg == null || (regSet.isStack(constraint) && !graph.nodes[vreg.varNum].use)) continue;
 					liveness[liveoutRow, vreg.varNum] = true;
-					addInterfere(vreg);
+					addInterfere(vreg, 0);
+					graph.nodes[vreg.varNum].color = true;
 				}
 				Overwrite(vreg, use, constraint)    => {
 					liveness[liveoutRow, vreg.varNum] = true;
-					addInterfere(vreg);
+					addInterfere(vreg, 0);
 				}
 				_ => ;
 			}
 			for (i < operands.length) match (operands[i]) {
 				Def(vreg, constraint) => {
+					if (vreg == null || (regSet.isStack(constraint) && !graph.nodes[vreg.varNum].use)) continue;
 					liveness[liveoutRow, vreg.varNum] = false;
 					graph.nodes[vreg.varNum].spillCost += curWeight;
+					if (opcode == ArchInstrs.ARCH_RELOAD) break;
+				}
+				Kill(constraint) => {
+					if (regSet.names[constraint] == "all") {
```

Comment:
We'll need to match on something other than the name for this check here. I don't understand why clearing the liveness is needed, TBH.

Reply:
I want to avoid reloading variables that are not going to be used between calls.